svgnative/ports/cairo/CairoSVGRenderer.cpp

```diff
@@ -0,0 +1,437 @@
+    // Cairo does not provide single API to draw "ellipse". See
+    // https://cairographics.org/cookbook/ellipses/
+
+    cairo_matrix_t  save_matrix;
+    cairo_get_matrix (mPath, &save_matrix); 
+
+    cairo_translate (mPath, cx, cy);
+    cairo_scale (mPath, rx, ry);
+    cairo_new_sub_path ( mPath );
+    cairo_arc (mPath, 0, 0, 1, 0, 2 * M_PI);
+
+    cairo_set_matrix ((cairo_t*)mPath, &save_matrix); 
+}
+
+void CairoSVGPath::MoveTo(float x, float y)
+{
+    cairo_move_to (mPath, x, y);
+    mCurrentX = x;
+    mCurrentY = y;
+}
+
+void CairoSVGPath::LineTo(float x, float y)
+{
+    cairo_line_to (mPath, x, y);
+    mCurrentX = x;
+    mCurrentY = y;
+}
+
+void CairoSVGPath::CurveTo(float x1, float y1, float x2, float y2, float x3, float y3)
+{
+    cairo_curve_to (mPath, x1, y1, x2, y2, x3, y3);
+    mCurrentX = x3;
+    mCurrentY = y3;
+}
+
+void CairoSVGPath::CurveToV(float x2, float y2, float x3, float y3)
+{
+    cairo_curve_to (mPath, mCurrentX, mCurrentY, x2, y2, x3, y3);
+    mCurrentX = x3;
+    mCurrentY = y3;
+}
+
+void CairoSVGPath::ClosePath() {
+    cairo_close_path (mPath);
+}
+
+CairoSVGTransform::CairoSVGTransform(float a, float b, float c, float d, float tx, float ty) {
+    cairo_matrix_init(&mMatrix, a, b, c, d, tx, ty);
+}
+
+void CairoSVGTransform::Set(float a, float b, float c, float d, float tx, float ty) {
+    // see CGSVGRenderer.cpp,
+    // internal of CGSVGTransform::CGSVGTransform() and CGSVGTransform::Set() are same.
+    cairo_matrix_init(&mMatrix, a, b, c, d, tx, ty);
+}
+
+void CairoSVGTransform::Rotate(float degree) {
+    cairo_matrix_rotate(&mMatrix, deg2rad(degree) );
+}
+
+void CairoSVGTransform::Translate(float tx, float ty) {
+    cairo_matrix_translate(&mMatrix, tx, ty);
+}
+
+void CairoSVGTransform::Scale(float sx, float sy) {
+    cairo_matrix_scale(&mMatrix, sx, sy);
+}
+
+void CairoSVGTransform::Concat(const Transform& other) {
+    cairo_matrix_t  result;
+    cairo_matrix_multiply(&result, &mMatrix, &(static_cast<const CairoSVGTransform&>(other).mMatrix));
+    // Cairo has no API to copy a matrix to another
+    cairo_matrix_init(&mMatrix, result.xx, result.yx, result.xy, result.yy, result.x0, result.y0); 
+}
+
+CairoSVGImageData::CairoSVGImageData(const std::string& base64, ImageEncoding encoding)
+{
+    std::string imageString = base64_decode(base64);
+    const unsigned char* blob_data = (const unsigned char*)malloc( imageString.size() );
+    if (!blob_data)
+        throw ("no memory\n");
+    memcpy((void *)blob_data, imageString.data(), imageString.size());
+
+    switch (encoding) {
+    case ImageEncoding::kJPEG:
+        mImageData = _cairo_image_surface_create_from_jpeg_stream( blob_data, imageString.size() );
+        if (!mImageData)
+            break;
+        cairo_surface_set_mime_data (mImageData, "image/jpeg", blob_data, imageString.size(), free, (void*)blob_data);
+        return;
+    case ImageEncoding::kPNG:
+        {
```

Comment:
Another indentation issue?
```cpp
case:
{
    block;
}
```
seems consistent with the rest of the code.

Reply:
Thanks, this strange blocking is caused by short-live temporal variable "png_closure" which is only used in the surface construction from PNG image data. I replace switch block by if-else block, for better readability.